tso-server/src/main/java/org/apache/omid/tso/CurrentTimestampOracleImpl.java

```diff
@@ -0,0 +1,171 @@
+
+    @VisibleForTesting
+    static class InMemoryTimestampStorage implements TimestampStorage {
+
+        long maxTime = 0;
+
+        @Override
+        public void updateMaxTimestamp(long previousMaxTime, long nextMaxTime) {
+            maxTime = nextMaxTime;
+            LOG.info("Updating max timestamp: (previous:{}, new:{})", previousMaxTime, nextMaxTime);
+        }
+
+        @Override
+        public long getMaxTimestamp() {
+            return maxTime;
+        }
+
+    }
+
+    private class AllocateTimestampBatchTask implements Runnable {
+        long previousMaxTime;
+
+        AllocateTimestampBatchTask(long previousMaxTime) {
+            this.previousMaxTime = previousMaxTime;
+        }
+
+        @Override
+        public void run() {
+            long newMaxTime = (System.currentTimeMillis() + TIMESTAMP_INTERVAL_MS) * MAX_TX_PER_MS;
+            try {
+                storage.updateMaxTimestamp(previousMaxTime, newMaxTime);
+                maxAllocatedTime = newMaxTime;
+                previousMaxTime = newMaxTime;
+            } catch (Throwable e) {
+                panicker.panic("Can't store the new max timestamp", e);
+            }
+        }
+    }
+
+    static final long MAX_TX_PER_MS = 1_000_000; // 1 million
+    private static final long TIMESTAMP_REMAINING_THRESHOLD = 3_000 * MAX_TX_PER_MS; // max number of transactions in 3 seconds
+    private static final long TIMESTAMP_INTERVAL_MS = 10_000; // 10 seconds interval
+
+    private long lastTimestamp;
+    private long maxTimestamp;
+
+    private TimestampStorage storage;
+    private Panicker panicker;
+
+    private long nextAllocationThreshold;
+    private volatile long maxAllocatedTime;
+
+    private Executor executor = Executors.newSingleThreadExecutor(
+            new ThreadFactoryBuilder().setNameFormat("ts-persist-%d").build());
+
+    private Runnable allocateTimestampsBatchTask;
+
+    @Inject
+    public CurrentTimestampOracleImpl(MetricsRegistry metrics,
+                               TimestampStorage tsStorage,
+                               Panicker panicker) throws IOException {
+
+        this.storage = tsStorage;
+        this.panicker = panicker;
+
+        metrics.gauge(name("tso", "maxTimestamp"), new Gauge<Long>() {
+            @Override
+            public Long getValue() {
+                return maxTimestamp;
+            }
+        });
+
+    }
+
+    @Override
+    public void initialize() throws IOException {
+
+        this.lastTimestamp = this.maxTimestamp = storage.getMaxTimestamp();
+
+        this.allocateTimestampsBatchTask = new AllocateTimestampBatchTask(lastTimestamp);
+
+        // Trigger first allocation of timestamps
+        executor.execute(allocateTimestampsBatchTask);
+
+        // Waiting for the current epoch to start. Occurs in case of fallback when the previous TSO allocated the current time frame.
+        while ((System.currentTimeMillis() * MAX_TX_PER_MS) < this.lastTimestamp);
```

Comment:
Busy wait - dangerous !! Please add short sleeps within the loop. 
Would you interrupt the process if the condition does not ever hold? 
Finally, please improve the comments - when you apply this code? Failover scenario?